### ISSUE TITLE
Three line fix for DANE TLSA DNS records going out of date every 90 days with LE

### ISSUE THREAD

**Issue by johnheenan** [virtualmin/virtualmin-gpl#803]
Free LE certificates are commonly used with Virtualmin. They need to be renewed every 90 days. DANE TLSA DNS records do not need to be updated provided:
1) The reissued LE certificates use `reuse-key` with certbot so the key is not changed.
2) The selector key of the TLSA record (second digit) is 1 (for public key) and not 0 (full certificate)

So 3-0-2 and 3-0-2 DANE TLS records should be avoided. Information at https://community.letsencrypt.org/t/please-avoid-3-0-1-and-3-0-2-dane-tlsa-records-with-le-certificates/7022

Virtualmin writes 3-0-1 records. Below is a fix to change to 3-1-1 records. I have confirmed this works by disabling domain TLSA DNS, reenabling it and testing from site https://dane.sys4.de/ (after a delay).

These code changes have nothing to do with setting up an addiitonal key TLSA DNS key for key rotation. If a key is changed then the TLSA DNS records need to be reissued (for example by disabling and enabling as above). There will be a small amount of time when records will be out of sync.

Check `--resuse-key` is in use from Webmin, Webmin, Webmin Configuration, Cog (for module configuration), Next, Re-use existing LE keys?, Yes, Save

Edit function `create_tlsa_dns_record` in file `webmin/virtual-server/feature-dns.pl`.

**Change lines:**
```
	"openssl x509 -in ".quotemeta($temp)." -outform DER 2>/dev/null | ".
	"openssl sha256 2>/dev/null");

```
**to:**
```
	"openssl x509 -in ".quotemeta($temp)." -pubkey -outform DER 2>/dev/null | ".
	"openssl rsa -pubin -outform DER 2> /dev/null | openssl sha256 2>/dev/null");

```
**Change line:**
```
	 'values' => [ 3, 0, 1, $1 ] };
```
**to**
```
	 'values' => [ 3, 1, 1, $1 ] };
```

**Comment by johnheenan** [comment]
So, why does this work? The whole idea of TLSA DNS is to gain independence from CAs, such as LE (Let's Encrypt) and others. The subject public key part of an x509 CA issued certificate is originally issued through the private key being certified and as long as the private key does not change, the subject public key does not change. The entire x509 certificate issued by a trusted CA that can be confirmed through the signature. But we don't need it.

So why not just extract the subject public key from the private key?

It is simpler but there are a number of problems with this approach:
```
openssl rsa -in ssl.key -pubout -outform DER | openssl sha256 
```
1) How do we know if rsa was used? x509 explicilty tells us the algorthm and the number of bits.
2) Virtualmin stores the location of the x509 file in its domain config, not the private key file ssl.key

**Comment by jcameron** [comment]
Does this only work for Let's Encrypt? We also need to support the case where the user uploads a cert from another CA ..

**Comment by johnheenan** [comment]
> Does this only work for Let's Encrypt?

I have not tried other CAs.

How far back does the DANE protocol go up the chain trust hierarchy before it refuses to trust a SMTP/TLS certificate issued by a CA and presented by a SMTP mail server? Does DANE cache results if it does go up a trust hierarchy?

It is up to intermediate and root CAs if they choose to publish TLSA DNS records for their certificates for DANE to check, if it chooses.

In principle I suppose DANE can and should work with self signed certificates. If so, then it does not matter which CA is used. With a 3 1 1 TLSA DNS record the CA issued certificate is not being authenticated, only the subject public key, which is included in the certificate and can be independently checked and verified with DNSSEC.

DANE is capable of saying "we don't care if you use self signed certs or CAs for SMTP/TLS if you use 3 1 1 TLSA DNS records"

But does DANE care? If it does care, do CA's care and so issue TLSA DNS records for their certificates?

How aboout giving Virtualmin users an option to generate either 3 0 1 certifcates (as occurs now) or generate 3 1 1 certifcates, until more is known. 

Another issue, even if DANE works with self signed certs and CAs other than LE, DANE survey sites may fault sites for using self signed certs or CAs that don't meet some additional standard they consider.

**Comment by johnheenan** [comment]
TL/DR to above: If someone reports that DANE works with self signed certicates then DANE should work fine with other CAs then LE.

**Comment by johnheenan** [comment]
These are my results changing a top level .com domain from using LE to self signed in Virtualmin

1) Four Virtualmin certificate files were updated (ssl.cert, ssl.combined, ssl.everything, ssl.key). File ssl.ca was not updated.
2) The TLSA DNS records in Virtualmin were updated
3) DNSSEC validation site https://dnsviz.net was happy with a re-analysis
4) DANE SMTP Validator site https://dane.sys4.de/ was happy with revalidation.
5) DNSSEC and DNS DANE TLS adoption survey site at https://stats.dnssec-tools.org/explore/ takes several days to update.

So my conclusion is that any CA will work. As to whether the survey site is happy. I don't know yet.

**Comment by jcameron** [comment]
So the current code is actually adding both the cert and the CA to the input file to the command `openssl x509 -in ".quotemeta($temp)." -outform DER 2>/dev/null ` , as seen here :

https://github.com/virtualmin/virtualmin-gpl/blob/master/feature-dns.pl#L4709

Would an even simpler fix to be not add the CA cert to that file, and just generate a 3 1 1 record?

**Comment by johnheenan** [comment]
The file referenced is the ssl.cert file and it only contains one certificate: the single x509 certificate from the CA, which is just a signed envelope of the original subject public key for the domain and other infomation.

Currently the entire x509 certifcate file DER output is hashed to produce the 3 0 1 TLSA DNS record.

DANE does not need this to authenticate. DANE is happy to just work with a hash of the subject public key to form a 3 1 1 TLSA DNS record. It can get this from the x509 certifcate from the CA in ssl.cert OR from the private key in ssl.key. I showed both methods above.

**Comment by johnheenan** [comment]
THere is another issue. Both methods require the algorthm to be known.

**Comment by johnheenan** [comment]
When DANE wants to validate SMTP domain authenticty it examines DNS to see if there is a TLSA DNS record for the service. If so it examines if the record is a 3 0 [12] type or 3 1 [12] type.

If it is a 3 0 1 type then DANE hashes the DER output from entire x509 certifcate provided by the smtp server and compares.

If it is a 3 1 1 type then DANE hashes the DER output from the extracted subject public key included in the x509 certifcate provided by the smtp server and compares.

It is up to DANE if it want to go further up the hierarchy and validate signer certifcates. It looks like it does not care to do so for standard DANE. I am waiting to see if the DANE survey site is more strict.

**Comment by chris001** [comment]
DANE doesn't need to go up the DNS hierarchy, because the all the way up the DNS hierarchy is already protected by DNSSEC.  It's impossible for a bad actor to slip his counterfeit certificate's hash into the DNS records without breaking DNSSEC signatures, causing the domain to fail to resolve thru DNS, red flag, ideally a monitor would notify the domain owner with an instant notification saying Signature Failed on a DNS record.

**Comment by johnheenan** [comment]
I meant the certifcate chain hierarchy, not the DNS hierarchy. 

Below is an image from the DANE survey site for a domain I control:
![image](https://github.com/virtualmin/virtualmin-gpl/assets/22131790/9b5e344e-07df-47bd-9e9c-309d41cf09a4)

Examining the second record, the survey site has gone up the certifcate hierarchy, generated the hash for CA certificate with name R3 (which is Let's Encrypt) and appears to have matched it with a DNS TLSA record (from where?). The CA certificate is valid for 10 years up to 2025.

For self-signed certifcates the hierarchy is from mail subdomain to main domain. I am waiting for a new report for a self signed certficate from the DANE survey site to see what it says.

I suppose if one wants to know if a CA is compatible with DANE then it should be sufficient for the CA to state it is, meaning DANE can verify their certifcates through DNSSEC.

As to whether DANE does or not check chain certifcates or can be confgured to or not, I don't know, All  I can say is that the DANE survey site appears to be checking.

**Comment by johnheenan** [comment]
I know the SMTP server will present two packaged certifcates from a single file: a CA cert and a domain cert signed by the CA. But why does the DANE survey site choose not to ignore the CA cert?

**Comment by johnheenan** [comment]
With regard to specific issues raised for Virtualmin, my view is that until more is known and trialled, Virtualmin should offer a choice between 3 0 1 type TLSA DNS records (as now) and 3 1 1 records, with a default to 3 1 1 records since it will solve known problems for most use cases (using LE).

Adding the following to lillustrate DANE is not just 'bye bye CAs' for SMTP.

Verisign is a CA. Following is from the end of a 2015 Verisign Blog https://blog.verisign.com/security/how-dane-strengthens-security-for-tls-smime-and-other-applications/

> In short, DANE provides the ability to use DNSSEC to perform the critically important function of secure key learning and verification. It can use the DNS directly to distribute and authenticate certificates and keys for endpoints. It can also work in conjunction with today’s public CA system by applying additional constraints about which CAs are authorized to issue certificates for specific services or users – thereby significantly reducing risks in the currently deployed CA system.

Trust in CAs is not high for good reason. It look likes Verisign sees DANE as been able to pick and choose which CAs to use for which services, if any. 

In addition, who would trust any specific organisation to self manage their own certifcation with self signing?

In addition with DNSSEC, which DANE relies upon, ultimate trust lies with a single registrar DS record. This is arguably worse than trusting a CA and is a massive point of failure that can cripple functionality.

**Comment by johnheenan** [comment]
Getting some of the mystery cleared up (I hope). Unfortunately, there is no clear source of information and I am basing this on interpretation of a cryptic post at https://community.letsencrypt.org/t/dane-and-upcoming-le-issuer-certs/134172, particularly this statement (but in context):

> TL;DR If you're using DANE-TA(2) (certificate usage 2) TLSA records with Let's Encrypt cert chains, you need to augment the TLSA RRSet with additional digests for the upcoming "R3" and/or "E1" issuer CAs.

My impression is  3 1 1 TLSA DNS records are suitable for self signed certs and CA signed certs but do not take any included CA cert into consideration for anything.

To require a CA cert to be taken into consideration then the TLSA DNA record must start with 2 instead of 3. In this case the CA cert must be included with the CA signed cert AND the CA cert TLSA DNS hash digest records must be published in the same zone file as for the domain TLSA DNS records.

RRSet records are just DNSSEC signatures for other DNS records.

**Comment by chris001** [comment]
These might provide some answers:

[SMTP Security by Opportunistic DANE TLS](https://datatracker.ietf.org/doc/html/rfc7672).
[The DANE protocol: Updates and Operator Guidance](https://datatracker.ietf.org/doc/html/rfc7671).
[Using DANE TLSA records with SRV records](https://datatracker.ietf.org/doc/html/rfc7673).

**Comment by johnheenan** [comment]
> These might provide some answers:
> 
> [SMTP Security by Opportunistic DANE TLS](https://datatracker.ietf.org/doc/html/rfc7672). 
[The DANE protocol: Updates and Operator Guidance](https://datatracker.ietf.org/doc/html/rfc7671). 
[Using DANE TLSA records with SRV records](https://datatracker.ietf.org/doc/html/rfc7673).

Thanks @chris001. 

These are interesting from RFC7671

(3) or (2) below refers to value of first digit in DNS TLSA record.

Start of section 5.2
> This section updates [RFC6698] by specifying a new operational
>    requirement for servers publishing TLSA records with a usage of
>    DANE-TA(2): such servers MUST include the TA certificate in their TLS
>    server certificate message unless all such TLSA records are "2 0 0"
>    records that publish the server certificate in full.
> 

Ah ha! Looks like I got it right in post above!

End of section 5.1
> While DANE-EE(3) TLSA records are expected to be by far the most
>    prevalent, as explained in Section 5.2, DANE-TA(2) records are a
>    valid alternative for sites with many DANE services.  Note, however,
>    that virtual hosting is more complex with DANE-TA(2).  Also, with
>    DANE-TA(2), server operators MUST ensure that the server is
>    configured with a sufficiently complete certificate chain and need to
>    remember to replace certificates prior to their expiration dates.

I am pleased that Virtualmin is capable of easily implementing DANE-TA(2)

DANE-TA(2) appears to be DNSSEC hash digest verified of two certificates: (CA cert) + (CA signed domain cert) with both certifcates made available by a service (such as SMTP).

Start of section 5.1

> Authentication via certificate usage DANE-EE(3) TLSA records involves
>    simply checking that the server's leaf certificate matches the TLSA
>    record.  In particular, the binding of the server public key to its
>    name is based entirely on the TLSA record association.  The server
>    MUST be considered authenticated even if none of the names in the
>    certificate match the client's reference identity for the server.
>    This simplifies the operation of servers that host multiple Customer
>    Domains, as a single certificate can be associated with multiple
>    domains without having to match each of the corresponding reference
>    identifiers.
> 

So no mention of a CA here.

The following is not relevant, unless you find the terminology exasperating (99% of us?)

My experience of RFCs is that they are incomprehensible rushed post conference after thoughts that only make sense to those who have been working together on a code base and agree a code implementation should drive a standard. That is an RFC standard does not drive original code and is not a request for comment (RFC). They are comments on extracted code comments by those who believe their code implementation is up to a standard. Hence thay have all the pain of discovered corner cases and are all about viewing trees, when we want is to view the forest. Hence, a big issue with RFCs is unclear and assumed teminology that lacks definition, usage of language that never became mainstream and later attempts with new RFCs to reconcile with an invented industry language that everyone imagined was in an RFC but wasn't. Although with DANE, there is no big industry to have evolved its own language.

**Comment by johnheenan** [comment]
> I am waiting for a new report for a self signed certficate from the DANE survey site to see what it says.

The DANE survey site results for a site with a self signed cert is in. The DANE survey site is happy. Survey site is https://stats.dnssec-tools.org/explore/

The only difference is that the SMTP/TLS certicates tab only showed one certifcate instead of two, as noted above for a LE signed domain certifcate.

Out of curiosity, I examined the Virtualminpostfix `/etc/postfix/sni_map` file. All line entries include entries for ssl.key and and ssl.cert files. Most also include an entry for a ssl.ca file. Some lines for domains that used LE were missing an entry for ssl.ca. I don't know if this is important, just noting it

The entry for the tested self signed domain did not list a ssl.ca file.

Maybe it is best to assume that entries in the SMTP/TLS certicates tab of the DANE survey site do nothing more than form digest hash versions of any certifcates offered to them by the SMTP server, independent of whether an offered CA files is used by DANE: yes for TA(2),  no for EE(3).

**Comment by johnheenan** [comment]
Some recent comments (Sept 2022) from https://community.letsencrypt.org/t/understanding-smtp-dane-implementation-options/184274

> Let's ping [@ietf-dane](https://community.letsencrypt.org/u/ietf-dane) here, who is an expert in this field.

Quotes from response at https://community.letsencrypt.org/t/understanding-smtp-dane-implementation-options/184274/4:
...

> I recommend against attempting to use the root CA public key hash as a stable fire and forget TLSA record. Even the root CA used by Let's Encrypt will eventually change, and you yourself might stop using Let's Encrypt, ...

...

> So to the question of what is best practice. It is indeed "3 1 1 + 3 1 1", where you configure Let's Encrypt to not change your key during regular certificate renewals (set reuse_key = true in the renewal .conf file), and arrange to inject a new key for certbot to use, only after that key's hash has been been published in DNS for at least a few TTLs in advance.
> 
> That way, if you forget to update the TLSA records, the certificate renewal process just keeps using the same key that already works, you keep obtaining new certs (for all those non-DANE clients to verify), and everything just works...
> 

The main point I take to take from this is that moving from a 3 0 1 DNS hash digest record to a 3 1 1 record, as proposed with the three lne fix, is best practice.

Why? Because for 3 1 1 record, DANE ignores any signing certificate (self signed, CA, up to date or not). With a 3 1 1 record DANE only pays attention to the original key (the subject publice key part), which does not need to change when a key is resigned.

DANE is used for authnticating SMTP for non relay clients. An email client can choose to ignore DANE and pass a message back that a signing cert is out of date or is not trusted. The three line fix does not alter any existing practices in this regard.

**Comment by Lawkss** [comment]
`Verdict:
At least one of your mail server domains does not have an active DANE scheme for a reliable rollover of certificate keys.

Technical details:
Mail server (MX)	DANE rollover scheme
mail.vom-bruch.com.	no
Test explanation:
We check if there is an active scheme with at least two DANE TLSA records to reliably handle certificate rollovers on your receiving mail servers (MX).

Such a scheme will be proven useful when there is a need to update your mail server certificate(s). It can prevent that DANE becomes invalid during the transition period which could endanger mail deliverability at your domain. A rollover scheme could but does not need to be 'active' all the time.

We recommend you to apply one of the following two schemes with double DANE TLSA records:

Current + Next ("3 1 1" + "3 1 1"): Publish two "DANE-EE(3) SPKI(1) SHA2-256(1)" records, one for the current and one for the next TLS certificate of your mail server.
Current + Issuer CA ("3 1 1" + "2 1 1"): Publish a "DANE-EE(3) SPKI(1) SHA2-256(1)" record for the current TLS certificate of your mail server, and also a "DANE-TA(2) SPKI(1) SHA2-256(1)" record for the current root or intermediate certificate of the (not necessarily public) certificate authority.
The test also accepts other combinations of DANE TLSA records i.e. "3 x x" + "3 x x" or "3 x x" + "2 x x". However other types than the above will generally be more error prone and less interoperable.`

Any chance we will get the rollover scheme implemented? <3

**Comment by johnheenan** [comment]
Rollover, as in post above, is more icing on the cake to make a potenetial problem of a few minutes go away and is a distraction from the real issue.

Since I put my three line code fix in place at the top (a move from 3 0 1 to 3 1 1), I have so far had no further problems. No rollover certificate, just a straight one for one replacement, better than no repacement, which is the real problem.

**Comment by johnheenan** [comment]
As long as the private key does not change and `--resuse-key` is used (see top) then a simple rollover might be just to leave the previous DNS TLSA record in place with the new one alongside. The start and end dates will overlap. The previous to the current previous DNS TLSA record can be deleted, if it exists.

**Comment by johnheenan** [comment]
Is rollover needed? I don't think it is for DANE use if a '3 1 1' scheme is in use and the private key does not change.

The reason for this is that for any reissued TLS certificate, the external certification is not used for the DNS TLSA record with a '3 1 1' scheme. Only the private key certification of the 'subject name' (the DNS name in practice) is used.

Since the previous private key is used with `--reuse-key` and a '3 1 1' scheme, no rollover is required. It does not matter if any DNS query finds a new or previous DNS TLSA record, as long as the dates overlap.

Of course if the private key does change or if the DANE scheme consults the full TLS certificate with a '3 0 1' scheme then rollover is best practice. One of the main points of DNS TLSA is to get away from excessive reliance on externally certified TLS certificates. That is, DNS allows uses of self certification.

**Comment by johnheenan** [comment]
Adding for completeness.

After making edits to `webmin/virtual-server/feature-dns.pl`, as shown in the top post of this issue, the following is required:

1. Restart webmin, such as from command `/etc/webmin/restart`
2. Regenerate TLSA records. One way is 'DNS Settings', 'DNS Options', 'TLSA records enabled' radio button No, 'Save', 'TLSA records enabled' radio button Yes, 'Save'

**Comment by Capitanguex** [comment]
I make this changes and it works good till the next update of Webmin. Is it not possible to integrate this in for example Webmin, Webmin, Webmin Configuration, Cog (for module configuration), Next, Re-use existing LE keys? For example a possibility here to select the scheme 3 0 1 or 3 1 1 what sets the correct code in feature-dns.pl and check also when updated ? Thanks

**Comment by jcameron** [comment]
We could potentially make this `3 1 1` mode an option, for users who prefer it? If the only change is what certs are included in the TLSA signature, this should be easy to implement.

**Comment by johnheenan** [comment]
Why not make 3 1 1 the default option?

Advantages of 3 1 1 over 3 0 1:
1. Full self certification with no dependence on a CA for a completely valid trust chain.
2. It is hard to disagree 3 1 1 is best practice.
3. It keeps getting hammered into us to preferably use 3 1 1 with those polite non solicited emails telling us we are doing it wrong.

**Comment by jcameron** [comment]
I'll do some more research and update this ticket with details.

Is there *any* case where `3 1 1` is not the right choice? Like for self-signed certs or certs from traditional issuers?

**Comment by johnheenan** [comment]
> Is there _any_ case where `3 1 1` is not the right choice? Like for self-signed certs or certs from traditional issuers?

I am not an expert. It is a security issue where the word 'right' is less important than managing 'risk' and what is practical.

For self signed certs `3 1 1` looks like the only solution. For CA signed certs from traditional issuers, it can be argued is `3 1 1` provides a practical safeguard against TLS certificates going out of date or getting withdrawn due to a CA incident.

For the 99.9% of relying on Let's Encrypt renewals every three months, I cannot see any practical benefit to raising the requirement for 311 to 301 for a TLSA RR. In fact 311 provided a safeguard to allow continued use of DANE when a TLS certificate is out of date.

I don't want to engage in theoretical security drama discussions when one of the single biggest ongoing practical security issues is lack of valid DANE records for use in email, risking MITM attacks. If Postfix is using Opportunistic DANE TLS to relay email and it finds invalid DANE records then it automatically downgrades. Who is going to use Postfix with mandatory DANE TLS and risk repeated tempfails? Reference https://www.postfix.org/TLS_README.html#client_tls_dane

For a TLSA RR certificate usage specified by 3 (DANE-EE), where the certificate does not need to be signed by a third party but might be, then it does not make practical sense for 99.9% of to use a selector of 0 (match entire signed certificate), rather than 1 (just match against public key of certificate).

**Not directly related but relevant as their is a bug in the current 301 system as is:**

For 311 TLSA RR, a TLS certificate renewal and a DNS TLSA RR renewal don't need to be synchronised, unless the TLS public key is reissued (such as when a TLS certificate is reissued rather than renewed).

So, if Virtualmin is going to allow 301 TLSA RR then they also need to reissue DNS TLSA records when a TLS certificate is reissued to ensure the TLSA records remain valid.

**Comment by johnheenan** [comment]
> So, if Virtualmin is going to allow 301 TLSA RR then they also need to reissue DNS TLSA records when a TLS certificate is reissued to ensure the TLSA records remain valid.

Sorry, shoud read instread:

So, if Virtualmin is going to allow 301 TLSA RR then they also need to **RENEW** DNS TLSA records when a TLS certificate is **RENEWED** to ensure the TLSA records remain valid.

This is not required when 311 records are used if a TLS certificate is only **RENEWED** and so is **NOT REISSUED**. Both TLS and DNS TLSA can keep happily renewing independently of each other.

**Comment by jcameron** [comment]
I hadn't heard of this RENEW record before - do you have a link to more details about it?

**Comment by johnheenan** [comment]
> I hadn't heard of this RENEW record before - do you have a link to more details about it?

I should use **RESIGN** instead of **RENEW**, such as when cron job `/etc/webmin/bind8/resign.pl` is run by Virtualmin

If a certbot (or other) job results in a Let's Encrypt (or other CA) signed TLS certificate renewal **AND** if a 301 DNS TLSA record depends on this certificate then there must be a forced DNS TLSA resign ahead of schedule to prevent loss of record  synchronisation. This is currently not done in Virtualmin.

For a 311 record no such forced resign is required **IF** the CA signed TLS certificate is only renewed and not reissued.

Some might be puzzled why a CA signed certificate is not required to prove a valid trusted chain. DNSSEC provides the trust chain. Cooperating Name server registrars allow DNSSEC records.

**Comment by jcameron** [comment]
Is this resign needed in `3 0 1` mode even when private key is left unchanged and the cert is just re-signed by the CA, such as via the `certbot` --resign flag?

**Comment by johnheenan** [comment]
> Is this resign needed in `3 0 1` mode even when private key is left unchanged and the cert is just re-signed by the CA, such as via the `certbot` --resign flag?

Yes resign is required,  because the CA signature has changed even though the public key (shown) and private key (not shown) are unchanged. 

For `3 1 1` mode the CA signature is not relevant.

It is worth pointing out that `3 1 1` mode just works for the default 10 year self signed cert, just regular resigning is required.

**Comment by jcameron** [comment]
Ok, work on an option to change the TLSA mode is in progress ...

**Comment by jcameron** [comment]
Ok, if anyone wants to try this, check out the attached patch. It adds a template option in the SSL section to use the new mode.

**Comment by iliajie** [comment]
@johnheenan You can apply the patch by running:

```
webmin patch https://github.com/virtualmin/virtualmin-gpl/commit/15d9854
```

**Comment by johnheenan** [comment]
I cannot find any visible options. 

Backed up the four affected files, applied patch, rebooted, cleared theme cache

No options visible:

![Image](https://github.com/user-attachments/assets/41d9fcbb-7a17-49be-ace5-49d6a3396ccf)

**Comment by iliajie** [comment]
This option should be checked in the templates:

<img width="1469" alt="Image" src="https://github.com/user-attachments/assets/ab148512-1bde-43dc-8954-4d424c047be9" />

**Comment by johnheenan** [comment]
Still cannot find relevant options in Virtualmin 7.30.8 in Server Templates or Virtualmin Configuration.

Your screenshot looks like a combination of SSL settings from Server Templates and Virtualmin Configuration

**From Server Templates:**

![Image](https://github.com/user-attachments/assets/bbc79c49-0f2c-4129-8fa2-4b6a6d63179d)

**From Virtualmin Configuration:**

![Image](https://github.com/user-attachments/assets/0a071611-5191-4491-818b-f56e3e1d8b45)

**Comment by iliajie** [comment]
Alright! In that case, you should manually edit `/etc/webmin/virtual-server/config` and add the line `tlsa_records=2`.

**Comment by jcameron** [comment]
@johnheenan did you apply the patch? From that template screenshot, it looks like it wasn't applied, or maybe Webmin wasn't restarted.

**Comment by iliajie** [comment]
@johnheenan, there have been quite a lot of changes, and the patch might not apply cleanly anymore. If you’re comfortable with it, you could try the [development build](https://software.virtualmin.dev/) of Virtualmin. Just be careful—it may still have some bugs.

**Comment by johnheenan** [comment]
Yes, from a diff with the four backed up files, the patch ignored two of the four files (feature-ssl.pl and lang/en).

I am reversing the patch now.

As well as running `/etc/webmin/restart`, I also rebooted and cleared the theme cache.

Thanks for the great work.

**Comment by johnheenan** [comment]
I found three patch reject files and the specific reasons for their rejection.

**patch reject for feature-dns.pl:**
```
cat  feature-dns.pl.rej
diff a/feature-dns.pl b/feature-dns.pl  (rejected hunks)
@@ -4892,12 +4903,12 @@ sub create_sshfp_dns_record
         'values' => [ $tn, 1, $1 ] };
 }

-# sync_domain_tlsa_records(&domain, [force-mode])
+# sync_domain_tlsa_records(&domain, [force-mode], [force-selector])
 # Replace all TLSA records for a domain with its actual SSL certs (if enabled)
 # force-mode 0 = use config, 1 = enable, 2 = disable
 sub sync_domain_tlsa_records
 {
-my ($d, $force) = @_;
+my ($d, $force, $force_sel) = @_;
 my $tmpl = &get_template($d->{'template'});
 &pre_records_change($d);
 my ($recs, $file) = &get_domain_dns_records_and_file($d);
```

Rejection reason: the following line is an addition to the original and so could not be matched:
```
my $tmpl = &get_template($d->{'template'});
```


**patch reject for feature-ssl.pl:**
```
cat  feature-ssl.pl.rej
diff a/feature-ssl.pl b/feature-ssl.pl  (rejected hunks)
@@ -3662,7 +3662,8 @@ sub show_template_ssl
        &hlink($text{'newweb_tlsa_records'}, "config_tlsa_records"),
        &ui_radio("ssl_tlsa_records", $tmpl->{'ssl_tlsa_records'},
          [ $tmpl->{'default'} ? ( ) : ( [ '', $text{'tmpl_default'} ] ),
-           [ 1, $text{'yes'} ],
+           [ 1, $text{'newweb_tlsa_records1'} ],
+           [ 2, $text{'newweb_tlsa_records2'} ],
            [ 0, $text{'no'} ] ]));
```
Rejection reason: no match anywhere for existing content, such as  `config_tlsa_records`, except in reject file:
```
grep -r config_tlsa_records
virtual-server/feature-ssl.pl.rej:      &hlink($text{'newweb_tlsa_records'}, "config_tlsa_records"),
```

**patch reject for lang/en:**

```
cat lang/en.rej
diff a/lang/en b/lang/en        (rejected hunks)
@@ -1972,6 +1972,8 @@ newweb_renew_letsencrypt_def=Default (7 days)
 newweb_erenew_letsencrypt=Days before expiry must be a number
 newweb_always_ssl=Generate SSL certificate for domains without SSL website?
 newweb_tlsa_records=Create TLSA DNS records for SSL certificates?
+newweb_tlsa_records1=Yes, with full certificate
+newweb_tlsa_records2=Yes, with just public key
 newweb_combined_cert=Configure Apache to use
 newweb_combined_cert2=Combined cert and CA file
 newweb_combined_cert1=Cert file only
```
Rejection reason: wrong file:
```
grep -r 'Create TLSA DNS records for SSL certificates'
virtual-server/config.info:tlsa_records=Create TLSA DNS records for SSL certificates?,1,1-Yes,0-No
virtual-server/lang/en.rej: newweb_tlsa_records=Create TLSA DNS records for SSL certificates?
virtual-server/help/config_tlsa_records.html:<header>Create TLSA DNS records for SSL certificates?</header>
```

**Comment by johnheenan** [comment]
> Thanks for the great work.

I am not trying to be ironic. The work has been done. It is a great advance and puts Virtualmin leagues ahead.

Currently it is just clean up work.

Can Virtualmin do DANE properly? Yes, soon
Anything else? No.

**Comment by jcameron** [comment]
One thing you could try if you want all the latest code and patches is our daily release dev versions, from https://software.virtualmin.dev/

However, I'd recommend testing on a non-production system first!

**Comment by Lawkss** [comment]
My cert renewal shaming mail from dnssec-tools.org that always rolls in after 90 days: 

Sometimes it also requires a BIND reload for some reason for it to be picked up faster. 

```
About the survey: https://stats.dnssec-tools.org/about.html

  This has become a recurring issue.  Any chance the TLSA configuration could
  be made more robust and be monitored to avoid future intermittent outages? ]

The TLSA RRsets of some of your email servers do not match their actual certificate chains.  Issue details for the affected domains:

    eloi.at vom-bruch.com

can be seen at:

    https://stats.dnssec-tools.org/explore/?eloi.at
    ...

The issues can be resolved by removing or updating the associated DNS DANE TLSA records.

    - "3 0 [12]" vs. Let's Encrypt:
      https://community.letsencrypt.org/t/please-avoid-3-0-1-and-3-0-2-dane-tlsa-records-with-le-certificates/7022/17

    - Best practice "3 1 1" rollover methodology:
      https://mail.sys4.de/pipermail/dane-users/2018-February/000440.html

    - Monitoring code snippet:
      https://list.sys4.de/hyperkitty/list/dane-users@list.sys4.de/thread/NKDBQABSTAAWLTHSZKC7P3HALF7VE5QY/

    - Realtime check of current DANE SMTP authentication:
      https://www.huque.com/bin/danecheck-smtp
```


I hope the patch works its way into production :)

**Comment by jcameron** [comment]
> I hope the 311 patch works its way into production :)

Oh yeah, it will be in the next release..

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.

**Comment by johnheenan** [comment]
I accept this issue has been fixed and I should have closed it. Thanks.

**Comment by iliaross** [comment]
Thank you for the feedback and closing it! Appreciate it!
